module/zfs/ddt_log.c

```diff
@@ -0,0 +1,750 @@
+	 * really correlated with the number of entries, but I don't think
+	 * that's reasonable to expect anyway.
+	 */
+	dmu_object_info_t doi;
+	uint64_t nblocks;
+	dmu_object_info(ddt->ddt_os, ddt->ddt_log_active->ddl_object, &doi);
+	nblocks = doi.doi_physical_blocks_512;
+	dmu_object_info(ddt->ddt_os, ddt->ddt_log_flushing->ddl_object, &doi);
+	nblocks += doi.doi_physical_blocks_512;
+
+	ddt_object_t *ddo = &ddt->ddt_log_stats;
+	ddo->ddo_count =
+	    avl_numnodes(&ddt->ddt_log_active->ddl_tree) +
+	    avl_numnodes(&ddt->ddt_log_flushing->ddl_tree);
+	ddo->ddo_mspace = ddo->ddo_count * sizeof (ddt_log_entry_t);
+	ddo->ddo_dspace = nblocks << 9;
+}
+
+void
+ddt_log_begin(ddt_t *ddt, size_t nentries, dmu_tx_t *tx, ddt_log_update_t *dlu)
+{
+	ASSERT3U(nentries, >, 0);
+	ASSERT3P(dlu->dlu_dbp, ==, NULL);
+
+	if (ddt->ddt_log_active->ddl_object == 0)
+		ddt_log_create(ddt, tx);
+
+	/*
+	 * We want to store as many entries as we can in a block, but never
+	 * split an entry across block boundaries.
+	 */
+	size_t reclen = P2ALIGN_TYPED(
+	    sizeof (ddt_log_record_t) + sizeof (ddt_log_record_entry_t) +
+	    DDT_PHYS_SIZE(ddt), sizeof (uint64_t), size_t);
+	ASSERT3U(reclen, <=, UINT16_MAX);
+	dlu->dlu_reclen = reclen;
+
+	uint64_t nblocks = howmany(nentries, DDT_LOG_BSIZE / dlu->dlu_reclen);
+	uint64_t offset = ddt->ddt_log_active->ddl_size;
+	uint64_t length = nblocks * DDT_LOG_BSIZE;
+
+	VERIFY0(dnode_hold(ddt->ddt_os, ddt->ddt_log_active->ddl_object, FTAG,
+	    &dlu->dlu_dn));
+	dnode_set_storage_type(dlu->dlu_dn, DMU_OT_DDT_ZAP);
+
+	VERIFY0(dmu_buf_hold_array_by_dnode(dlu->dlu_dn, offset, length,
+	    B_FALSE, FTAG, &dlu->dlu_ndbp, &dlu->dlu_dbp,
+	    DMU_READ_NO_PREFETCH));
+
+	dlu->dlu_tx = tx;
+	dlu->dlu_block = dlu->dlu_offset = 0;
+}
+
+static ddt_log_entry_t *
+ddt_log_alloc_entry(ddt_t *ddt)
+{
+	ddt_log_entry_t *ddle;
+
+	if (ddt->ddt_flags & DDT_FLAG_FLAT) {
+		ddle = kmem_cache_alloc(ddt_log_entry_flat_cache, KM_SLEEP);
+		memset(ddle, 0, DDT_LOG_ENTRY_FLAT_SIZE);
+	} else {
+		ddle = kmem_cache_alloc(ddt_log_entry_trad_cache, KM_SLEEP);
+		memset(ddle, 0, DDT_LOG_ENTRY_TRAD_SIZE);
+	}
+
+	return (ddle);
+}
+
+static void
+ddt_log_update_entry(ddt_t *ddt, ddt_log_t *ddl, ddt_lightweight_entry_t *ddlwe)
+{
+	/* Create the log tree entry from a live or stored entry */
+	avl_index_t where;
+	ddt_log_entry_t *ddle =
+	    avl_find(&ddl->ddl_tree, &ddlwe->ddlwe_key, &where);
+	if (ddle == NULL) {
+		ddle = ddt_log_alloc_entry(ddt);
+		ddle->ddle_key = ddlwe->ddlwe_key;
+		avl_insert(&ddl->ddl_tree, ddle, where);
+	}
+	for (int p = 0; p < ddlwe->ddlwe_nphys; p++)
+		ddle->ddle_phys[p] = ddlwe->ddlwe_phys[p];
+	ddle->ddle_type = ddlwe->ddlwe_type;
+	ddle->ddle_class = ddlwe->ddlwe_class;
+}
+
+void
+ddt_log_entry(ddt_t *ddt, ddt_lightweight_entry_t *ddlwe, ddt_log_update_t *dlu)
+{
+	ASSERT3U(dlu->dlu_dbp, !=, NULL);
+
+	ddt_log_update_entry(ddt, ddt->ddt_log_active, ddlwe);
```

Comment:
Thinking again, this would requite allowing both logs to be written in parallel, so it may be not so trivial, but the idea still feels interesting at least to think.

Reply:
By design, the flushing log can't be modified once flushing starts. Otherwise, it has the potential to grow unbounded, if updates are coming in faster than we can flush them.

Could we do something in-between? Unsure. This version was hard-won though, so we'd have to think about it a lot first.

Reply:
It would not be unbounded.  Since the logs are hash-indexed, as the write process moving ddl_checkpoint further probability of insertion into active log with proportionally decrease.

Reply:
I was thinking about this today, and I actually think this is possible, and quite brilliant I think. I'm not even sure how to describe concisely; is it a bit like a ring buffer? Or a kind of moving window over the two logs? Like, it's got some odd notion of sorting, but not exactly. It's cool, I get it, I like it, I just can't quite describe it yet :laughing:

As is my way, I think it should not be attempted yet. It feels like something that could be tricky to get right, and I don't really want to introduce new spanners just as I'm really starting to feel like what's here is stable. I'm pretty sure it doesn't require any kind of on-disk change to do it in the future; there's no inherent difference between the active and flushing logs, and hash order is maintained regardless of whether or not the flushing is growing.

Lets land this beast, and then have a swing at it when we're sure we've got the glitches ironed out and can measure any performance improvement this brings. Fair?

Reply:
Sure.  It is not a blocker for having this released.